NUT-XX and NUT-XX+1: Mint / Melt Bitcoin On-Chain (#107)
cashubtc/nuts: 17.md

NUT-17: Mint tokens Bitcoin On-Chain
==========================

`optional`

---

This NUT describes the process of minting tokens on Bitcoin On-Chain analog to NUT-04 for Bitcoin Lightning. 

# Mint quote

To request a mint quote, the wallet of `Alice` makes a `POST /v1/mint/quote/{method}` request where `method` is the payment method requested (here `btconchain`). 

```http
POST https://mint.host:3338/v1/mint/quote/btconchain
```

The wallet of `Alice` includes the following `PostMintQuoteBtcOnchainRequest` data in its request:

```json
{
  "amount": <int>,

>   "unit": <str_enum["sat"]>

Author: Why should we remove the unit-field from the PostMintQuoteOnchainRequest? If the mint uses blink as a backend for handling the onchain transactions it can support onchain snd/receive with usd as a currency. Hopefully in the future there will be similar services. https://dev.blink.sv/api/usd-onchain-send

>   "unit": <str_enum["sat"]>

Reviewer: This is a question of scoping this NUT. To support other units where a conversion rate is involved how expiration is handled as well as how confirmation time is defined is much more important as the conversion rate could change outside an acceptable slippage.  So in order to support other units how confirmation and expiration is defined needs to be included as well.

}
```
 with the requested `amount` and the `unit`. 
 
 The mint `Bob` then responds with a `PostMintQuoteBtcOnchainResponse`:

```json
{
  "quote": <str>,

>   "address": <str>,

Reviewer: What if we change this from an address to a [BIP-21 URI](https://github.com/bitcoin/bips/blob/master/bip-0021.mediawiki)? This would enable the mint to optionally support [BIP-78 payjoins](https://github.com/bitcoin/bips/blob/master/bip-0078.mediawiki).

Reviewer 2: IUC, BIP-21 is meant for clickable links, additional information such as comments, amount etc, should not be part of the `address` field as it would introduce ambiguity. I think it's fair to assume that the backend would be able to pay most common address formats, and should error (for melt quote) if it doesn't support a particular address format.

  "paid": <bool>,

>   "expiry": <int>

Reviewer 2: This expiry field might be problematic (cf @gandlafbtc's comment).

> ```json
> {
>   "quote": <str>,
>   "address": <str>,
>   "paid": <bool>,
>   "expiry": <int>
> }
> ```

Reviewer: ```suggestion
```json
{
  "quote": <str>,
  "address": <str>,
  "state": <str_enum[STATE]>,
}
```
```

> Where `quote` is the quote ID and `address` is the payment request to fulfill. `paid` indicates whether the request as been paid and `expiry` is the Unix timestamp until which the mint quote is valid.
> 
> Note: `quote` is a **unique and random** id generated by the mint to internally look up the payment state. `quote` **MUST** remain a secret between user and mint and **MUST NOT** be derivable from the payment request. A third party who knows the `quote` ID can front-run and steal the tokens that this operation mints.

Reviewer: ```suggestion
Where `quote` is the quote ID and `address` is the payment request to fulfill. 

`state` is an enum string field with possible values `"UNPAID"`, `"PAID"`, `"PENDING"`, `"ISSUED"`:
- `"UNPAID"` means that the quote's request has not been paid yet.
- `"PAID"` means that the request has been paid.
- `"PENDING"` means that the quote is currently being issued.
- `"ISSUED"` means that the quote has already been issued.

Note: `quote` is a **unique and random** id generated by the mint to internally look up the payment state. `quote` **MUST** remain a secret between user and mint and **MUST NOT** be derivable from the payment request. A third party who knows the `quote` ID can front-run and steal the tokens that this operation mints.
```

Author: What is the difference between issued and pending? Is issued the initial state?

## Example

Request of `Alice` with curl:

```bash
curl -X POST https://mint.host:3338/v1/mint/quote/btconchain -d '{"amount": 10, "unit": "sat"}' -H "Content-Type: application/json"
```

Response of `Bob`:

> 
> ```json
> {
>   "quote": "DSGLX9kevM...",
>   "address": "bc1qkyfgd7mus7ykfd7qkwakq75qsf7rtm...",
>   "paid": false,
>   "expiry": 1701704757
> }
> ```

Reviewer: ```suggestion

```json
{
  "quote": "DSGLX9kevM...",
  "address": "bc1qkyfgd7mus7ykfd7qkwakq75qsf7rtm...",
  "paid": UNPAID,
}
```
```

      "C_": "0224f1c4c564230ad3d96c5033efdc425582397a5a7691d600202732edc6d4b1ec"
    },
    {
      "id": "009a1f293253e41e",
      "amount": 8,
      "C_": "0277d1de806ed177007e5b94a8139343b6382e472c752a74e99949d511f7194f6c"
    }
  ]
}
```

## Unblinding signatures

Upon receiving the `BlindSignatures` from the mint `Bob`, the wallet of `Alice` unblinds them to generate `Proofs` (using the blinding factor `r` and the mint's public key `K`, see BDHKE [NUT-00][00]). The wallet then stores these `Proofs` in its database:

```json
[
  {
  "id": "009a1f293253e41e",
  "amount": 2,
  "secret": "407915bc212be61a77e3e6d2aeb4c727980bda51cd06a6afc29e2861768a7837",
  "C": "02bc9097997d81afb2cc7346b5e4345a9346bd2a506eb7958598a72f0cf85163ea"
  },
  {
  "id": "009a1f293253e41e",
  "amount": 8,
  "secret": "fe15109314e61d7756b0f8ee0f23a624acaa3f4e042f61433c728c7057b931be",
  "C": "029e8e5050b890a7d6c0968db16bc1d5d5fa040ea1de284f6ec69d61299f671059"
  }
]
```

## Settings
The settings for this nut indicate the supported method-unit pairs for minting and whether minting is supported or not. They are part of the info response of the mint ([NUT-06][06]) which in this case reads 
```json
{

>   "14": {

Reviewer: ```suggestion
  "17": {
```

    "supported": true,
    "methods": [
      {

>         "payment_method": "btconchain",

Reviewer 2: Should be `method`
```suggestion
        "method": "btconchain",
```

        "unit": "sat",

>         "min_amount": 1000,

Reviewer 2: Might be better to use a realistic minimal amount for on-chain here.
```suggestion
        "min_amount": 10000,
```
